staging/src/k8s.io/code-generator/cmd/conversion-gen/generators/conversion_test.go

```diff
@@ -0,0 +1,181 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generators
+
+import (
+	"reflect"
+	"testing"
+)
+
+func Test_extractTag(t *testing.T) {
+	testCases := []struct {
+		comments []string
+		expect   []string
+	}{
+		{
+			comments: []string{},
+			expect:   nil,
+		},
+		{
+			comments: []string{
+				"+k8s:conversion-gen=k8s.io/kubernetes/runtime.Object",
+			},
+			expect: []string{
+				"k8s.io/kubernetes/runtime.Object",
+			},
+		},
+		{
+			comments: []string{
+				"+k8s:conversion-gen=k8s.io/kubernetes/runtime.Object",
+				"+k8s:conversion-gen=k8s.io/kubernetes/runtime.List",
+			},
+			expect: []string{
+				"k8s.io/kubernetes/runtime.Object",
+				"k8s.io/kubernetes/runtime.List",
+			},
+		},
+		{
+			comments: []string{
+				"+k8s:conversion-gen=k8s.io/kubernetes/runtime.Object",
+				"+k8s:conversion-gen=k8s.io/kubernetes/runtime.Object",
+			},
+			expect: []string{
+				"k8s.io/kubernetes/runtime.Object",
+				"k8s.io/kubernetes/runtime.Object",
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		r := extractTag(tc.comments)
+		if r == nil && tc.expect != nil {
+			t.Errorf("case[%d]: expected non-nil", i)
+		}
+		if r != nil && tc.expect == nil {
+			t.Errorf("case[%d]: expected nil, got %v", i, r)
+		}
+		if r != nil && !reflect.DeepEqual(r, tc.expect) {
+			t.Errorf("case[%d]: expected %v, got %v", i, tc.expect, r)
+		}
+	}
+}
+
+func Test_extractExplicitFromTag(t *testing.T) {
+	testCases := []struct {
+		comments []string
+		expect   []string
+	}{
+		{
+			comments: []string{},
+			expect:   nil,
+		},
+		{
+			comments: []string{
+				"+k8s:conversion-gen:explicit-from=k8s.io/kubernetes/runtime.Object",
+			},
+			expect: []string{
+				"k8s.io/kubernetes/runtime.Object",
+			},
+		},
+		{
+			comments: []string{
+				"+k8s:conversion-gen:explicit-from=k8s.io/kubernetes/runtime.Object",
+				"+k8s:conversion-gen:explicit-from=k8s.io/kubernetes/runtime.List",
+			},
+			expect: []string{
+				"k8s.io/kubernetes/runtime.Object",
+				"k8s.io/kubernetes/runtime.List",
+			},
+		},
+		{
+			comments: []string{
+				"+k8s:conversion-gen:explicit-from=k8s.io/kubernetes/runtime.Object",
+				"+k8s:conversion-gen:explicit-from=k8s.io/kubernetes/runtime.Object",
+			},
+			expect: []string{
+				"k8s.io/kubernetes/runtime.Object",
+				"k8s.io/kubernetes/runtime.Object",
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		r := extractExplicitFromTag(tc.comments)
+		if r == nil && tc.expect != nil {
+			t.Errorf("case[%d]: expected non-nil", i)
+		}
+		if r != nil && tc.expect == nil {
+			t.Errorf("case[%d]: expected nil, got %v", i, r)
+		}
+		if r != nil && !reflect.DeepEqual(r, tc.expect) {
```

Comment:
Use `if diff := cmp.Diff(...); diff != "" { t.Errorf("Unexpected result (-want, +got):\n%s", diff) }` - or some helper lib like stretchr/testify which does this for you.

Reply:
corrected